File: serialize-and-deserialize-n-ary-tree/shaun-gy-bfs.cpp
/*
// Definition for a Node.
class Node {
public:
    int val = NULL;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/
/*

grandyang idea, cutting: bfs, need to know the number of children of each node
runtime 236ms, faster than 57%, mem less than 100%
*/

class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(Node* root) {
        if (!root) return string("#");
        string res;
        queue<Node*> q;
        q.push(root);
        while (!q.empty()) {
            Node *t = q.front(); q.pop();
            res += to_string(t->val) + " " + to_string(t->children.size()) + " ";
            for (Node *child : t->children) { //child won't be NULL if any child exists
                q.push(child);
            }
        }
        return res;
    }

    // Decodes your encoded data to tree.
    Node* deserialize(string data) {
        istringstream iss(data);
        queue<Node*> qNode;
        queue<int> qSize;
        string val = "", size = "";
        iss >> val;
        if (val == "#") return NULL;
        iss >> size;
        Node *root = new Node(stoi(val), {});
        qNode.push(root);
        qSize.push(stoi(size));
        
        while (!qNode.empty()) {
            Node *t = qNode.front(); qNode.pop();
            int len = qSize.front(); qSize.pop();
            //to fill t's children, len children
            for (int i = 0; i < len; ++i) {
                if (!(iss >> val)) break;
                if (!(iss >> size)) break;
                Node* cur = new Node(stoi(val), {}); //val won't be #
                qNode.push(cur);
                qSize.push(stoi(size));
                t->children.push_back(cur);
            }
        }
        return root;
    }
};


// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));

File: serialize-and-deserialize-binary-tree/shaun-gy-bfs.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 
 
 shaun bfs idea from c++ to java
 using array of string split to store tokens
 error prone, then use iterator for cleaner code, 
 
 Runtime: 18 ms, faster than 36.08% of Java online submissions for Serialize and Deserialize Binary Tree.
Memory Usage: 41.8 MB, less than 19.05% of Java online submissions for Serialize and Deserialize Binary Tree.
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        StringBuilder out = new StringBuilder();
        Queue<TreeNode> q = new LinkedList<>();
        if (root != null) q.offer(root);
        while (!q.isEmpty()) {
            TreeNode t = q.poll();
            if (t!=null) {
                out.append(t.val + " ");
                q.offer(t.left);
                q.offer(t.right);
            } else {
                out.append("# ");
            }
        }
        return out.toString();
    }
    
    // Decodes your encoded data to tree.
    //cleaner code using iterator
    public TreeNode deserialize(String data) {
        if (data.length() ==0) return null; //NULL okay too
        
        Queue<TreeNode> q = new LinkedList<>();
        //Iterator<String> it = Arrays.stream(data.split(" ")).iterator();
        Iterator<String> it = Arrays.asList(data.split(" ")).iterator();
        String token = it.next();
        TreeNode root = new TreeNode(Integer.valueOf(token));
        TreeNode cur = root;
        q.offer(root);
        
        while (!q.isEmpty()) {
            
            TreeNode node = q.poll();
            for (int i = 0; i < 2 &&it.hasNext(); i++) {
                token = it.next();
                if (!token.equals("#")) {
                    cur = new TreeNode(Integer.valueOf(token));
                    q.offer(cur);
                    switch(i){
                        case 0: node.left = cur; break;
                        case 1: node.right = cur; break;
                    }
                    
                }
            }
        }
        return root;

    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize2(String data) {
        if (data.length() ==0) return null; //NULL okay too
        
        Queue<TreeNode> q = new LinkedList<>();
        String [] a = data.split(" ");
        int i = 0;
        TreeNode root = new TreeNode(Integer.valueOf(a[i]));
        TreeNode cur = root;
        q.offer(root);
        i++;
        while (!q.isEmpty()) {
            
            if (i == a.length) break;
            String token = a[i];
            //if (token.length() == 0) continue;
            
            TreeNode node = q.poll();
            if (!token.equals("#")) {
                cur = new TreeNode(Integer.valueOf(token));
                q.offer(cur);
                node.left = cur;
            }
            i++;
            if (i == a.length) break;
            token = a[i];
            //if (token.length() == 0) continue;
            if (!token.equals("#")) {
                cur = new TreeNode(Integer.valueOf(token));
                q.offer(cur);
                node.right = cur;
            }
            i++;
        }
        return root;

    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.deserialize(codec.serialize(root));
